packages/framework/aqueduct/src/data-object-factories/dataObjectFactory.ts

```diff
@@ -32,28 +35,58 @@ export class DataObjectFactory<
 	TObj extends DataObject<I>,
 	I extends DataObjectTypes = DataObjectTypes,
 > extends PureDataObjectFactory<TObj, I> {
+	/**
+	 * @Remarks Use the props object based constructor instead.
+	 * No new features will be added to this constructor,
+	 * and it will eventually be deprecated and removed.
+	 */
 	public constructor(
 		type: string,
 		ctor: new (props: IDataObjectProps<I>) => TObj,
-		sharedObjects: readonly IChannelFactory[] = [],
-		optionalProviders: FluidObjectSymbolProvider<I["OptionalProviders"]>,
+		sharedObjects?: readonly IChannelFactory[],
+		optionalProviders?: FluidObjectSymbolProvider<I["OptionalProviders"]>,
 		registryEntries?: NamedFluidDataStoreRegistryEntries,
-		runtimeFactory: typeof FluidDataStoreRuntime = FluidDataStoreRuntime,
+		runtimeFactory?: typeof FluidDataStoreRuntime,
+	);
+	public constructor(props: DataObjectFactoryProps<TObj, I>);
+	public constructor(
+		propsOrType: DataObjectFactoryProps<TObj, I> | string,
+		maybeCtor?: new (doProps: IDataObjectProps<I>) => TObj,
+		maybeSharedObjects?: readonly IChannelFactory[],
+		maybeOptionalProviders?: FluidObjectSymbolProvider<I["OptionalProviders"]>,
+		maybeRegistryEntries?: NamedFluidDataStoreRegistryEntries,
+		maybeRuntimeFactory?: typeof FluidDataStoreRuntime,
 	) {
-		const mergedObjects = [...sharedObjects];
+		const props: DataObjectFactoryProps<TObj, I> =
+			typeof propsOrType === "string"
+				? {
+						type: propsOrType,
+						// both the arg and props base constructor require this param
+						// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+						ctor: maybeCtor!,
```

Comment:
Too bad TS doesn't do narrowing based on overloads (right?)

Reply:
it might technically be possible, but the typing would be very complicated, which kind of defeats the purpose in my opinion.